### PR TITLE
fix(engine): fix unit tests

### PR DESCRIPTION
**Pull request by SylvestreG**

### CODE REVIEW

File: modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh [hunk 1]
```diff
@@ -217,10 +216,40 @@ namespace         modules {
...
-        if (!svc)
+
+        std::pair<uint64_t, uint64_t> found_id(get_host_and_service_id(
+          name, description));
+        std::unordered_map<std::pair<uint64_t, uint64_t>,
+                           std::shared_ptr<service> >::const_iterator
+          found(configuration::applier::state::instance().services().find(found_id));
+
+        if (found != configuration::applier::state::instance().services().end() ||
+          found->second.get() != nullptr)
```
Comment (bouda1) [inline modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh]: Isn't it wrong?
`if (!svc)` becomes `found->second.get() != nullptr` ...
I think you wanted` !found->second`

File: modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh [hunk 2]
```diff
@@ -217,10 +216,40 @@ namespace         modules {
...
-        service* svc(::find_service(name, description));
-        if (!svc)
+
+        std::pair<uint64_t, uint64_t> found_id(get_host_and_service_id(
+          name, description));
+        std::unordered_map<std::pair<uint64_t, uint64_t>,
+                           std::shared_ptr<service> >::const_iterator
+          found(configuration::applier::state::instance().services().find(found_id));
+
+        if (found != configuration::applier::state::instance().services().end() ||
```
Comment (bouda1) [inline modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh]: !svc so we did not found it.
did you mean `found == configuration::applier::state::instance().services().end()`

File: modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh [hunk 3]
```diff
@@ -233,12 +262,41 @@ namespace         modules {
...
-        service* svc(::find_service(name, description));
-        if (!svc)
+        std::pair<uint64_t, uint64_t> found_id(get_host_and_service_id(
+          name, description));
+        std::unordered_map<std::pair<uint64_t, uint64_t>,
+                           std::shared_ptr<service> >::const_iterator
+          found(configuration::applier::state::instance().services().find(found_id));
+
+        if (found != configuration::applier::state::instance().services().end() ||
+          found->second.get() != nullptr)
```
Comment (bouda1) [inline modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh]: The same here. We did not found the service....

File: modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh [hunk 4]
```diff
@@ -233,12 +262,41 @@ namespace         modules {
...
+
+        char* name(my_strtok(args, ";"));
+        char* description(my_strtok(NULL, ";"));
+        std::pair<uint64_t, uint64_t> found_id(get_host_and_service_id(
+          name, description));
+        std::unordered_map<std::pair<uint64_t, uint64_t>,
+                           std::shared_ptr<service> >::const_iterator
+          found(configuration::applier::state::instance().services().find(found_id));
+
+        if (found != configuration::applier::state::instance().services().end() ||
```
Comment (bouda1) [inline modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh]: Here, I don't have the original code but it looks similar...

File: modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh [hunk 5]
```diff
@@ -249,17 +307,18 @@ namespace         modules {
 
         char* group_name(my_strtok(args, ";"));
         servicegroup_map::const_iterator sg_it{servicegroup::servicegroups.find(group_name)};
-        if (sg_it == servicegroup::servicegroups.end())
+        if (sg_it == servicegroup::servicegroups.end() ||
+          sg_it->second.get() == nullptr)
```
Comment (bouda1) [inline modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh]: boolean operator works on shared_ptr, `sg_it->second.get() == nullptr` is the same as `!sg_it->second`. :-p

File: modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh [hunk 6]
```diff
@@ -249,17 +307,18 @@ namespace         modules {
...
-               end(group->members.end());
-             it != end;
-             ++it)
-          if (it->second.get())
-            (*fptr)(it->second.get());
+               it2(sg_it->second->members.begin()),
+               end2(sg_it->second->members.end());
+             it2 != end2;
+             ++it2)
+          if (it2->second.get())
```
Comment (bouda1) [inline modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh]: get() is not useful.

File: modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh [hunk 7]
```diff
@@ -272,21 +331,21 @@ namespace         modules {
 
         char* group_name(my_strtok(args, ";"));
         servicegroup_map::const_iterator sg_it{servicegroup::servicegroups.find(group_name)};
-        if (sg_it == servicegroup::servicegroups.end())
+        if (sg_it == servicegroup::servicegroups.end() || sg_it->second.get() == nullptr)
```
Comment (bouda1) [inline modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh]: just !sg_it->second is sufficient.

File: modules/external_commands/src/compatibility.cc [hunk 8]
```diff
@@ -1059,33 +1058,33 @@ int process_hostgroup_command(int cmd,
              end2(it->second->services.end());
            it2 != end2;
            ++it2) {
-        if ((temp_service = it2->second.get()) == nullptr)
+        if (it2->second.get() == nullptr)
```
Comment (bouda1) [inline modules/external_commands/src/compatibility.cc]: to simplify

File: modules/external_commands/src/compatibility.cc [hunk 9]
```diff
@@ -1155,12 +1153,12 @@ int process_host_command(int cmd,
            end(temp_host->services.end());
          it != end;
          ++it) {
-      if ((temp_service = it->second.get()) == nullptr)
+      if (it->second.get() == nullptr)
```
Comment (bouda1) [inline modules/external_commands/src/compatibility.cc]: can be simplified

File: modules/external_commands/src/compatibility.cc [hunk 10]
```diff
@@ -1171,12 +1169,12 @@ int process_host_command(int cmd,
...
```
Comment (bouda1) [inline modules/external_commands/src/compatibility.cc]: idem

File: modules/external_commands/src/compatibility.cc [hunk 11]
```diff
@@ -1268,62 +1265,67 @@ int process_service_command(int cmd,
     return ERROR;
 
   /* find the service */
-  if ((temp_service = find_service(host_name, svc_description)) == nullptr)
+  std::pair<uint64_t, uint64_t> id(get_host_and_service_id(
+    host_name, svc_description));
+  std::unordered_map<std::pair<uint64_t, uint64_t>,
+                     std::shared_ptr<service> >::const_iterator
+    found(state::instance().services().find(id));
+  if (found == state::instance().services().end() || found->second.get() == nullptr)
```
Comment (bouda1) [inline modules/external_commands/src/compatibility.cc]: idem

File: modules/external_commands/src/compatibility.cc [hunk 12]
```diff
@@ -1355,22 +1357,19 @@ int process_servicegroup_command(int cmd,
...
   /* get the servicegroup name */
   if ((servicegroup_name = my_strtok(args, ";")) == nullptr)
     return ERROR;
 
   /* find the servicegroup */
-  servicegroup_map::const_iterator sg_it{servicegroup::servicegroups.find(servicegroup_name)};
-  if (sg_it == servicegroup::servicegroups.end())
+  servicegroup_map::const_iterator
+    sg_it{servicegroup::servicegroups.find(servicegroup_name)};
+  if (sg_it == servicegroup::servicegroups.end() || sg_it->second.get() == nullptr)
```
Comment (bouda1) [inline modules/external_commands/src/compatibility.cc]: idem

File: modules/external_commands/src/compatibility.cc [hunk 13]
```diff
@@ -1383,40 +1382,43 @@ int process_servicegroup_command(int cmd,
...
 
-      temp_service = find_service(it->first.first.c_str(),
-        it->first.second.c_str());
-      if (temp_service == nullptr)
+      std::pair<uint64_t, uint64_t> id(get_host_and_service_id(
+        it->first.first, it->first.second));
+      std::unordered_map<std::pair<uint64_t, uint64_t>,
+                         std::shared_ptr<service> >::const_iterator
+        found(state::instance().services().find(id));
+      if (found == state::instance().services().end() || found->second.get() == nullptr)
```
Comment (bouda1) [inline modules/external_commands/src/compatibility.cc]: idem

File: modules/external_commands/src/processing.cc [hunk 14]
```diff
@@ -380,7 +377,7 @@ void processing::_wrapper_enable_host_svc_notifications(host* hst) {
        it != end;
        ++it)
     if (it->second.get())
```
Comment (bouda1) [inline modules/external_commands/src/processing.cc]: get() is not needed

File: src/config.cc [hunk 15]
```diff
@@ -1237,10 +1236,15 @@ int check_servicegroup(servicegroup* sg, int* w, int* e) {
...
-    if (!temp_service) {
+
+    std::pair<uint64_t, uint64_t> id(get_host_and_service_id(
+      it->first.first, it->first.second));
+    std::unordered_map<std::pair<uint64_t, uint64_t>,
+      std::shared_ptr<service> >::const_iterator
+        found(state::instance().services().find(id));
+
+    if (found == state::instance().services().end() ||
+      found->second.get() == nullptr) {
```
Comment (bouda1) [inline src/config.cc]: the same remark

File: src/configuration/applier/contact.cc [hunk 16]
```diff
@@ -393,11 +393,10 @@ void applier::contact::resolve_object(
 
   // Find contact.
   contact_map::const_iterator ct_it{engine::contact::contacts.find(obj.contact_name())};
-  if (ct_it == engine::contact::contacts.end())
+  if (ct_it == engine::contact::contacts.end() || ct_it->second.get() == nullptr)
```
Comment (bouda1) [inline src/configuration/applier/contact.cc]: idem

File: src/configuration/applier/service.cc [hunk 17]
```diff
@@ -569,7 +569,7 @@ void applier::service::resolve_object(
     &deleter::objectlist);
 
   // Find host and adjust its counters.
-  umap<unsigned long, std::shared_ptr<com::centreon::engine::host>>::iterator
+  umap<unsigned long, std::shared_ptr<engine::host>>::iterator
```
Comment (bouda1) [inline src/configuration/applier/service.cc]: uint64_t for an id

File: src/downtimes/service_downtime.cc [hunk 18]
```diff
@@ -144,15 +147,19 @@ void service_downtime::print(std::ostream& os) const {
...
-  service* svc;
+  std::pair<uint64_t, uint64_t> id(get_host_and_service_id(
+    get_hostname(), get_service_description()));
+  std::unordered_map<std::pair<uint64_t, uint64_t>,
+                     std::shared_ptr<service> >::const_iterator
+    found(state::instance().services().find(id));
 
   /* find the host or service associated with this downtime */
-  if ((svc = ::find_service(get_hostname().c_str(), get_service_description().c_str())) == nullptr)
+  if (found == state::instance().services().end() || found->second.get() == nullptr)
```
Comment (bouda1) [inline src/downtimes/service_downtime.cc]: idem

File: src/downtimes/service_downtime.cc [hunk 19]
```diff
@@ -308,15 +318,20 @@ int service_downtime::subscribe() {
...
 
+  std::pair<uint64_t, uint64_t> id(get_host_and_service_id(
+    get_hostname(), get_service_description()));
+  std::unordered_map<std::pair<uint64_t, uint64_t>,
+                     std::shared_ptr<service> >::const_iterator
+    found(state::instance().services().find(id));
+
   /* find the host or service associated with this downtime */
-  if ((svc = ::find_service(get_hostname().c_str(), get_service_description().c_str())) == nullptr)
+  if (found == state::instance().services().end() || found->second.get() == nullptr)
```
Comment (bouda1) [inline src/downtimes/service_downtime.cc]: idem

File: src/macros.cc [hunk 20]
```diff
@@ -210,20 +210,27 @@ int grab_custom_macro_value_r(
...
-                            mac->host_ptr ? mac->host_ptr->get_name().c_str() : nullptr,
-                            arg2))) {
+
+      std::pair<uint64_t, uint64_t> id(get_host_and_service_id(
+        mac->host_ptr ? mac->host_ptr->get_name(): "", arg2));
+      std::unordered_map<std::pair<uint64_t, uint64_t>,
+                         std::shared_ptr<service> >::const_iterator
+        found(state::instance().services().find(id));
+
+      if (found != state::instance().services().end() ||
```
Comment (bouda1) [inline src/macros.cc]: && not ||

File: src/macros.cc [hunk 21]
```diff
@@ -210,20 +210,27 @@ int grab_custom_macro_value_r(
...
                    macro_name + 8,
-                   temp_service->custom_variables,
+                   found->second->custom_variables,
                    output);
       }
       /* else we have a service macro with a servicegroup name and a delimiter... */
       else {
         servicegroup_map::const_iterator sg_it{servicegroup::servicegroups.find(arg1)};
-        if (sg_it == servicegroup::servicegroups.end())
+        if (sg_it == servicegroup::servicegroups.end() || sg_it->second.get() == nullptr)
```
Comment (bouda1) [inline src/macros.cc]: get() not needed, in fact` !sg_it->second` is shorter.

File: src/macros/grab_value.cc [hunk 22]
```diff
@@ -229,15 +229,20 @@ static int handle_service_macro(
...
       else if (arg2) {
-        com::centreon::engine::service* svc(find_service(mac->host_ptr->get_name().c_str(), arg2));
-        if (!svc)
+        std::pair<uint64_t, uint64_t> id(get_host_and_service_id(
+          mac->host_ptr->get_name(), arg2));
+        std::unordered_map<std::pair<uint64_t, uint64_t>,
+                           std::shared_ptr<service> >::const_iterator
+          found(state::instance().services().find(id));
+
+        if (found == state::instance().services().end() || found->second.get() == nullptr)
```
Comment (bouda1) [inline src/macros/grab_value.cc]: idem

File: src/macros/grab_value.cc [hunk 23]
```diff
@@ -246,40 +251,44 @@ static int handle_service_macro(
...
       // On-demand macro with both host and service name.
-      com::centreon::engine::service* svc(find_service(arg1, arg2));
-      if (svc)
+      std::pair<uint64_t, uint64_t> id(get_host_and_service_id(
+        arg1, arg2));
+      std::unordered_map<std::pair<uint64_t, uint64_t>,
+                         std::shared_ptr<service> >::const_iterator
+        found(state::instance().services().find(id));
+
+      if (found == state::instance().services().end() || found->second.get() == nullptr)
```
Comment (bouda1) [inline src/macros/grab_value.cc]: idem

File: src/macros/grab_value.cc [hunk 24]
```diff
@@ -246,40 +251,44 @@ static int handle_service_macro(
...
-                   svc,
+                   found->second.get(),
                    output,
                    free_macro);
       // Else we have a service macro with a
       // servicegroup name and a delimiter...
       else {
         servicegroup_map::const_iterator sg_it{servicegroup::servicegroups.find(arg1)};
-        if (sg_it == servicegroup::servicegroups.end())
+        if (sg_it == servicegroup::servicegroups.end() || found->second.get() == nullptr )
```
Comment (bouda1) [inline src/macros/grab_value.cc]: idem

File: src/macros/grab_value.cc [hunk 25]
```diff
@@ -342,7 +351,7 @@ static int handle_servicegroup_macro(
   // Use the saved servicegroup pointer
   // or find the servicegroup for on-demand macros.
   servicegroup_map::const_iterator sg_it{servicegroup::servicegroups.find(arg1)};
-  if (sg_it == servicegroup::servicegroups.end())
+  if (sg_it == servicegroup::servicegroups.end() || sg_it->second.get() == nullptr)
```
Comment (bouda1) [inline src/macros/grab_value.cc]: idem

File: src/configuration/applier/service.cc (hunk 17 above)
Comment (SylvestreG) [reply src/configuration/applier/service.cc]: done

File: modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh (hunk 1 above)
Comment (SylvestreG) [reply modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh]: done

File: modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh (hunk 2 above)
Comment (SylvestreG) [reply modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh]: done

File: modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh (hunk 3 above)
Comment (SylvestreG) [reply modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh]: done

File: modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh (hunk 4 above)
Comment (SylvestreG) [reply modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh]: done

File: modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh (hunk 5 above)
Comment (SylvestreG) [reply modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh]: done

File: modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh (hunk 6 above)
Comment (SylvestreG) [reply modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh]: done

File: modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh (hunk 7 above)
Comment (SylvestreG) [reply modules/external_commands/inc/com/centreon/engine/modules/external_commands/processing.hh]: done

File: modules/external_commands/src/compatibility.cc (hunk 8 above)
Comment (SylvestreG) [reply modules/external_commands/src/compatibility.cc]: done

File: modules/external_commands/src/compatibility.cc (hunk 10 above)
Comment (SylvestreG) [reply modules/external_commands/src/compatibility.cc]: done

File: modules/external_commands/src/compatibility.cc (hunk 12 above)
Comment (SylvestreG) [reply modules/external_commands/src/compatibility.cc]: done

File: modules/external_commands/src/compatibility.cc (hunk 13 above)
Comment (SylvestreG) [reply modules/external_commands/src/compatibility.cc]: done

File: src/config.cc (hunk 15 above)
Comment (SylvestreG) [reply src/config.cc]: done

File: src/downtimes/service_downtime.cc (hunk 18 above)
Comment (SylvestreG) [reply src/downtimes/service_downtime.cc]: done

File: src/downtimes/service_downtime.cc (hunk 19 above)
Comment (SylvestreG) [reply src/downtimes/service_downtime.cc]: done

File: src/macros.cc (hunk 21 above)
Comment (SylvestreG) [reply src/macros.cc]: done

File: src/macros.cc (hunk 20 above)
Comment (SylvestreG) [reply src/macros.cc]: done

File: src/macros/grab_value.cc (hunk 22 above)
Comment (SylvestreG) [reply src/macros/grab_value.cc]: done

File: src/macros/grab_value.cc (hunk 23 above)
Comment (SylvestreG) [reply src/macros/grab_value.cc]: done

File: src/macros/grab_value.cc (hunk 25 above)
Comment (SylvestreG) [reply src/macros/grab_value.cc]: done

File: src/macros/grab_value.cc (hunk 24 above)
Comment (SylvestreG) [reply src/macros/grab_value.cc]: done

**Pull request merged**